### ISSUE TITLE
False negative when narrowing attribute type in subclass

### ISSUE THREAD

**Issue by LeeeeT** [python/mypy#12942]
**Bug Report**

Mypy doesn't complain about redefining an attribute in a subclass with a narrower type. I expect the following code to produce a static type checking error:

```python
class A:
    f: object

class B(A):
    f: float
```

Since this will fail:

```python
b = B()

a: A = b
a.f = "abc"

b.f.is_integer()
```

**Your Environment**

- Mypy version used: `0.960`
- Mypy configuration options: `strict=true`
- Python version used: `3.10`

**Comment by hauntsaninja** [comment]
Thanks, duplicate of #3208
